### PR TITLE
Upgrade: v11.2.0

### PR DESCRIPTION
**Pull request by kataras**
- Add `context#FullRequestURI` and `NewConditionalHandler` as requested at: #1167 and #1170 by @zheeeng;
- Add warning messages when an end-dev missuses the MVC dependencies registration and controllers handling;
- Add an example to execute the `ctx.Translate` (using the `i18n.Translate` from `/iris/middleware/i18n`) inside the templates: `{{call .tr "hi" "arguments..."}}` when `.tr` is registered as a view data `"tr": ctx.Translate`, end-dev can use a view-data wrapper for common-used functions as they already do;
- Websocket: A lot of cleanup and performance improvements that can theoretically solve the _C10M Problem_
- Add [tutorial](https://github.com/kataras/iris/tree/4cfdc6418532f49ff0159045a06080db0457eb55/_examples/tutorial/mongodb) for the Official MongoDB Go Driver.
- Add `Context#ResetRequest` and use it to update the request on `core/handlerconv.FromStdWithNext` as requested at #1180
- Do not call a controller's method if second output value of a dynamic dependency returned a not nil error as requested at #1187
- Give the ability to modify the whole session's cookie on `Start` and `Update/ShiftExpiration` methods, add a `StartWithPath` helper as requested at #1186
- Add `Context.ResponseWriter.IsHijacked` to report whether the underline connection is hijacked
- Add the ability to intercept the default error handler by seting a custom `hero.ErrorHandler` to MVC application-level or per controller as requested at[#1244](https://github.com/kataras/iris/issues/1244), example at: https://github.com/kataras/iris/blob/v11.2.0/_examples/mvc/error-handler/main.go
- Extract the `Delim` for redis sessiondb as requested at #1256
- Fix hero/mvc when map, struct, slice return `nil` as `null` in JSON responses, as reported at #1273
- Fix pongo2 addons can't be registered, as reported at #1284 and `view/django` pongo2 conversions are now type aliases
- Full re-write of the websocket module;
The v11.2.0 will contain a breaking change on the websockets usage, there is no other way to improve it. That one has its own repository for testing and experimentation to implement a faster and even more higher level websocket server and client written in Go which developers will be able select between [gorilla/websocket](https://github.com/gorilla/websocket) and [gobwas/ws](https://github.com/gobwas/ws) for the underline protocol communication and eventually will live into Iris, click [here](https://github.com/kataras/neffos) to watch the progress of the new Iris websocket capabilities and features if you can't wait more.
- Replace all static methods with `iris.Fileserver` package-level function and Party and APIBuilder method `HandleDir`, read more at https://github.com/kataras/iris/issues/1283
- The `ctx.HandlerName()` returns the current handler's function name of the chain. The **new** `ctx.RouteName()` returns the current route's name which is executing the handler. Also **new** `ctx.HandlerFileName()` which returns the exact program's source code position of the handler's function. 
- Redis backend service changed from "redigo" to "radix" which is more active project and gives more options for connection pooling and clustering. The old `/sessiondb/redis/service.Config` was replaced by `/sessiondb/redis.Config` (no need to import other path for custom configuration now). Also a `redis.Config.MaxActive` allows to customize the max number of connections and the old `redis.Config.IdleTimeout` was replaced by `redis.Config.Timeout` which sets timeouts for redis reads, writes and dialing. **Fix** a bug which removed the whole session (like `sess.Destroy()`  when using `sess.Clear()` with redis sessiondb.
- Implement a **new View Engine** for the [jet](https://github.com/CloudyKit/jet/) template parser as requested at: https://github.com/kataras/iris/issues/1281. Examples available at: [_examples/view/template_jet_0](https://github.com/kataras/iris/tree/v11.2.0/_examples/view/template_jet_0) and [_examples/view/template_jet_1_embedded](https://github.com/kataras/iris/tree/v11.2.0/_examples/view/template_jet_1_embedded)
- Add `mvc#Application.HandleWebsocket` to register websocket controllers, its methods as websocket events now automatically, add a new `stateless` struct field tag as well.
- Add `mvc#Before/AfterActivation.HandleMany and GetRoutes` methods as requested at: https://github.com/kataras/iris/issues/1292
- Implement a websocket stack exchange feature that allows scaling-out of websocket servers, add a built-in `websocket.NewRedisStackExchange` for scaling-out through redis pubsub system as requested at: https://github.com/kataras/neffos/issues/3.
- Fix `iris.WithoutBodyConsumptionOnUnmarshal` not be respected on `ctx.ReadForm` and `ctx.FormValues` as reported at: #1297
- **New feature**: Automatic Public domains for development using `iris.WithTunneling` configurator. Read more at: https://github.com/kataras/iris/issues/1305

More to come....stay tuned!

### CODE REVIEW

**Comment by jjhesk** [comment]
@kataras why sync map doesnt solve the problem?

**Comment by kataras** [comment]
You told me that it doesn't at our e-mail contact...I though it was buggy so I change it to our old-school way. Both ways should be safe, we already test it on running products and never got a datarace for about 2 years... I will need your code to test your case, please send me over the whole project and how you test it.

**Comment by speedwheel** [comment]
@kataras This example with the new mongodb driver came just in time. 

I've been using Iris + MongoDb extensively over the past 1.5 years in my projects. But I never used the official golang driver since it was still in alpha. Now that it's in beta, I might make the switch for the next project, also I am curious to test its performance comapred wit the [community version](https://github.com/globalsign/mgo) driver and see if there is any noticeable difference.

**Comment by jjhesk** [comment]
@kataras here i just applied the new code `server.go` and i discovered more race errors than before.. please hold on this patch.

**Comment by speedwheel** [comment]
After doing some tests it seems like the community drivers are 2x faster than the official ones for selecting a single document.

10 vs. 20k req/second

mongodb-new-vs-old

**Comment by kataras** [comment]
@speedwheel That's awesome man!! I saw from its code that it may require more allocations to do things than the mgo driver but I myself, didn't expect so big difference between those two, should I add a tutorial for the mgo driver as well? It will be indetical through, or should I leave it to you? 🥇

**Comment by kataras** [comment]
@jjhesk I know but I wanted to see if you get a difference data race stackrace, I replied to your new issue as well about byte buffer. Of course, the release is posponding until we fix **all the opened bugs by the community**, this includes the websocket datarace you mention, but still I can't re-produce that, can you make some code changes in your local iris and do further tests? It seems that you are deeply working on this for more than a week, you can open a PR or you could tell me what you learnt so far so I can continue investigating it. Thank you a lot, once again!

File: context/context.go [hunk 1]
```diff
@@ -114,6 +114,18 @@ type Context interface {
...
 	Request() *http.Request
+	// ResetRequest sets the Context's Request,
+	// It is useful to store the new request created by a std *http.Request#WithContext() into Iris' Context.
+	// Use `ResetRequest` when for some reason you want to make a full
+	// override of the *http.Request.
+	// Note that: when you just want to change one of each fields you can use the Request() which returns a pointer to Request,
+	// so the changes will have affect without a full override.
+	// Usage: you use a native http handler which uses the standard "context" package
+	// to get values instead of the Iris' Context#Values():
+	// r := c.Request()
```
Comment (kataras) [inline context/context.go]: typo fixed with: https://github.com/kataras/iris/pull/1175/commits/1a8d162da3c263939f618aa335f4086101f8be7e

**Comment by speedwheel** [comment]
@kataras I opened an issue on jira about this to create more awareness regarding the performance side.

Yes I will do the tutorial, it might come in handy for some users,  as for me, I will still continue using mgo for now.

**Comment by jjhesk** [comment]
tested locally with socket updated version and there are 2 discovered race detections. 


==================
WARNING: DATA RACE
Write at 0x00c001d0a680 by goroutine 32:
  runtime.slicecopy()
      /usr/local/Cellar/go/1.11.4/libexec/src/runtime/slice.go:221 +0x0
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*messageSerializer).serialize()
      /Users/_local_/go/src/github.com/valyala/bytebufferpool/bytebuffer.go:73 +0x17b
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*emitter).Emit()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/emitter.go:37 +0xa6
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*connection).Emit()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/connection.go:579 +0x1cc
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.EmitToAllSubbed()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/ws_core.go:405 +0x23d
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.loopMonitorBB.func1()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/bd_bigbang.go:39 +0x3ed
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.CoreLoopEngineV2()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/bd_core.go:31 +0x4fe
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.loopMonitorBB()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/bd_bigbang.go:12 +0xe0

Previous read at 0x00c001d0a680 by goroutine 799:
  runtime.slicecopy()
      /usr/local/Cellar/go/1.11.4/libexec/src/runtime/slice.go:221 +0x0
  compress/flate.(*compressor).fillStore()
      /usr/local/Cellar/go/1.11.4/libexec/src/compress/flate/deflate.go:521 +0xe6
  compress/flate.(*compressor).write()
      /usr/local/Cellar/go/1.11.4/libexec/src/compress/flate/deflate.go:552 +0xf8
  compress/flate.(*Writer).Write()
      /usr/local/Cellar/go/1.11.4/libexec/src/compress/flate/deflate.go:709 +0x5c
  github.com/gorilla/websocket.(*flateWriteWrapper).Write()
      /Users/_local_/go/src/github.com/gorilla/websocket/compression.go:102 +0x85
  github.com/gorilla/websocket.(*Conn).WriteMessage()
      /Users/_local_/go/src/github.com/gorilla/websocket/conn.go:760 +0xe5
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*connection).Write()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/connection.go:354 +0xd2
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*connection).writeDefault()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/connection.go:366 +0x77
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*Server).emitMessage()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/server.go:355 +0x2d2
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*emitter).EmitMessage()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/emitter.go:32 +0x127
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*emitter).Emit()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/emitter.go:41 +0x114
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*connection).Emit()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/connection.go:579 +0x1cc
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.StartWebsocket.func1.12.1()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/ws_core.go:371 +0x73
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.ExecSyncRountine()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/connectionLock.go:60 +0x74
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.StartWebsocket.func1.12()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/ws_core.go:370 +0x73

Goroutine 32 (running) created at:
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.StartWebsocket()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/ws_core.go:394 +0x336
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.SetupWebCombineClient()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/webhost.go:58 +0x76b
  main.startProc()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/main.go:57 +0x103
  main.main()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/main.go:36 +0x57

Goroutine 799 (running) created at:
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.StartWebsocket.func1()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/ws_core.go:367 +0xce1
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp.(*Server).Handler.func1()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/temp/server.go:103 +0xf4
  github.com/kataras/iris/context.DefaultNext()
      /Users/_local_/go/src/github.com/kataras/iris/context/context.go:1208 +0x134
  github.com/kataras/iris/context.(*context).Next()
      /Users/_local_/go/src/github.com/kataras/iris/context/context.go:1217 +0x5b
  _/Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x.SetupWebCombineClient.func2()
      /Users/_local_/Documents/bigbangsurvivalrun/backendc/main/core_x/webhost.go:31 +0x117
  github.com/kataras/iris/context.Do()
      /Users/_local_/go/src/github.com/kataras/iris/context/context.go:922 +0xa5
  github.com/kataras/iris/context.(*context).Do()
      /Users/_local_/go/src/github.com/kataras/iris/context/context.go:1094 +0x62
  github.com/kataras/iris/core/router.(*routerHandler).HandleRequest()
      /Users/_local_/go/src/github.com/kataras/iris/core/router/handler.go:227 +0x751
  github.com/kataras/iris/core/router.(*Router).BuildRouter.func1()
      /Users/_local_/go/src/github.com/kataras/iris/core/router/router.go:84 +0xc4
  github.com/kataras/iris/core/router.(*Router).ServeHTTP()
      /Users/_local_/go/src/github.com/kataras/iris/core/router/router.go:161 +0x69
  net/http.serverHandler.ServeHTTP()
      /usr/local/Cellar/go/1.11.4/libexec/src/net/http/server.go:2741 +0xc4
  net/http.(*conn).serve()
      /usr/local/Cellar/go/1.11.4/libexec/src/net/http/server.go:1847 +0x80a
==================

**Pull request merged**